src/ameba/rule/lint/unused_generic_or_union.cr

```diff
@@ -0,0 +1,78 @@
+module Ameba::Rule::Lint
+  # A rule that disallows unused generics or unions (`String?`, `StaticArray(Int32, 10)`, etc).
+  #
+  # For example, these are considered invalid:
+  #
+  # ```
+  # String?
+  #
+  # Float64 | StaticArray(Float64, 10)
+  #
+  # def size
+  #   Float64 | Int32
+  #   0.1
+  # end
+  # ```
+  #
+  # And these are considered valid:
+  #
+  # ```
+  # a : Int32 = 10
+  #
+  # klass = String?
+  #
+  # alias MyType = Float64 | StaticArray(Float64, 10)
+  #
+  # def size : Float64
+  #   0.1
+  # end
+  # ```
+  #
+  # YAML configuration example:
+  #
+  # ```
+  # Lint/UnusedGenericOrUnion:
+  #   Enabled: true
+  # ```
+  class UnusedGenericOrUnion < Base
+    properties do
+      since_version "1.7.0"
+      description "Disallows unused generics or unions"
+    end
+
+    GENERIC_MSG = "Generic is not used"
+    UNION_MSG   = "Union is not used"
+
+    def test(source : Source)
+      AST::ImplicitReturnVisitor.new(self, source)
+    end
+
+    def test(source, node : Crystal::Call, node_is_used : Bool)
+      return if node_is_used || !path_or_generic_union?(node)
+
+      issue_for node, UNION_MSG
+    end
+
+    def test(source, node : Crystal::Generic, node_is_used : Bool)
+      issue_for node, GENERIC_MSG unless node_is_used
+    end
+
+    private def path_or_generic_union?(node : Crystal::Call) : Bool
+      node.name == "|" && node.args.size == 1 && !!(obj = node.obj) &&
+        valid_type_node?(obj) && valid_type_node?(node.args.first)
+    end
+
+    private def valid_type_node?(node : Crystal::ASTNode) : Bool
+      case node
+      when Crystal::Path, Crystal::Generic, Crystal::Self, Crystal::TypeOf, Crystal::Underscore
```

Comment:
Shouldn't (`Crystal::...`) `SizeOf`, `InstanceSizeOf`, `AlignOf`, `InstanceAlignOf`, `OffsetOf`, `PointerOf` be included here as well?

Reply:
I didn't as those aren't allowed in type restrictions (which this is supposed to check). Those would fall under `Lint/UnusedPseudoMethod` imo

Reply:
Unless I'm missing something this rule ought to check for unused usages of generic types and unions, not type restrictions?

Reply:
That's correct - if i'm remembering correctly unions are only allowed inside type restrictions

Reply:
https://carc.in/#/r/hmnp

`String | Int32` outside of a type restriction is a method call on String that creates a union, that's what we're checking for here

Reply:
Those language constructs don't create unions though (see the carcin link above), so they should not be checked

Reply:
You're right. In such a case `TypeOf` shouldn't be included there either.

Reply:
My bad, that's correct.